org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ProcessTaskManager.java

```diff
@@ -46,159 +51,131 @@ public class ProcessTaskManager {
 		return t;
 	});
 
-public ProcessTaskManager(Compiler compiler, int startingIndex) {
-	this.compiler = compiler;
-	this.startingIndex = startingIndex;
+	public ProcessTaskManager(Compiler compiler, int startingIndex) {
+		this.compiler = compiler;
+		this.startingIndex = startingIndex;
 
-	this.currentIndex = 0;
-	this.availableIndex = 0;
-	this.size = PROCESSED_QUEUE_SIZE;
-	this.sleepCount = 0; // 0 is no one, +1 is the processing thread & -1 is the writing/main thread
-	this.units = new CompilationUnitDeclaration[this.size];
+		this.units = new ArrayBlockingQueue<>(PROCESSED_QUEUE_SIZE);
 
-	synchronized (this) {
-		this.processingTask = executor.submit(this::compile);
+		this.processingTask = executor.submit(this::processing);
 	}
-}
 
 // add unit to the queue - wait if no space is available
-private synchronized void addNextUnit(CompilationUnitDeclaration newElement) {
-	while (this.units[this.availableIndex] != null) {
-		//System.out.print('a');
-		//if (this.sleepCount < 0) throw new IllegalStateException(Integer.valueOf(this.sleepCount).toString());
-		this.sleepCount = 1;
+	private void addNextUnit(Object newElement) {
 		try {
-			wait(250);
-		} catch (InterruptedException ignore) {
-			// ignore
+			this.units.put(newElement);
+		} catch (InterruptedException interrupt) {
+			throw new RuntimeException(interrupt);
 		}
-		this.sleepCount = 0;
 	}
 
-	this.units[this.availableIndex++] = newElement;
-	if (this.availableIndex >= this.size)
-		this.availableIndex = 0;
-	if (this.sleepCount <= -1)
-		notify(); // wake up writing thread to accept next unit - could be the last one - must avoid deadlock
-}
+	private Object lastSignal = null;
```

Comment:
please don't init to defaults, this makes debugger always jump to every useless init line...

Reply:
ok